nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/SiteToSiteProvenanceReportingTask.java

```diff
@@ -0,0 +1,354 @@
+            getLogger().debug("No events to send because no events have been created yet.");
+            return;
+        }
+
+        if (firstEventId < 0) {
+            Map<String, String> state;
+            try {
+                state = context.getStateManager().getState(Scope.LOCAL).toMap();
+            } catch (IOException e) {
+                getLogger().error("Failed to get state at start up due to {}:"+e.getMessage(), e);
+                return;
+            }
+            if (state.containsKey(LAST_EVENT_ID_KEY)) {
+                firstEventId = Long.parseLong(state.get(LAST_EVENT_ID_KEY)) + 1;
+            }
+
+            if(currMaxId < firstEventId){
+                getLogger().debug("Current provenance max id is {} which is less than what was stored in state as the last queried event, which was {}. This means the provenance restarted its " +
+                        "ids. Restarting querying from the beginning.", new Object[]{currMaxId, firstEventId});
+                firstEventId = -1;
+            }
+        }
+
+        if (currMaxId == (firstEventId - 1)) {
+            getLogger().debug("No events to send due to the current max id being equal to the last id that was queried.");
+            return;
+        }
+
+        List<ProvenanceEventRecord> events;
+        try {
+            events = context.getEventAccess().getProvenanceEvents(firstEventId, context.getProperty(BATCH_SIZE).asInteger());
+        } catch (final IOException ioe) {
+            getLogger().error("Failed to retrieve Provenance Events from repository due to: " + ioe.getMessage(), ioe);
+            return;
+        }
+
+        if (events == null || events.isEmpty()) {
+            getLogger().debug("No events to send due to 'events' being null or empty.");
+            return;
+        }
+
+        final String nifiUrl = context.getProperty(INSTANCE_URL).evaluateAttributeExpressions().getValue();
+        URL url;
+        try {
+            url = new URL(nifiUrl);
+        } catch (final MalformedURLException e1) {
+            // already validated
+            throw new AssertionError();
+        }
+
+        final String hostname = url.getHost();
+        final String platform = context.getProperty(PLATFORM).evaluateAttributeExpressions().getValue();
+
+        final Map<String, ?> config = Collections.emptyMap();
+        final JsonBuilderFactory factory = Json.createBuilderFactory(config);
+        final JsonObjectBuilder builder = factory.createObjectBuilder();
+
+        while (events != null && !events.isEmpty()) {
+            final long start = System.nanoTime();
+
+            // Create a JSON array of all the events in the current batch
+            final JsonArrayBuilder arrayBuilder = factory.createArrayBuilder();
+            for (final ProvenanceEventRecord event : events) {
+                arrayBuilder.add(serialize(factory, builder, event, getComponentName(procGroupStatus, event), hostname, url, rootGroupName, platform));
```

Comment:
This can be pretty expensive potentially, continually iterating over every component in the flow, looking for the correct identifier. What do you think about just converting the ProcessGroupStatus into a Map<String, String> where the key is a component ID and the value is the component name? Then we could just do a quick get() on the map